crates/core/tedge_api/src/workflow/mod.rs

```diff
@@ -384,6 +417,63 @@ impl OperationAction {
 
         Ok(new_state)
     }
+
+    /// Rewrite a command state before pushing it to a builtin operation actor
+    ///
+    /// Depending the action is to trigger or await the operation,
+    /// set the status to schedule or executing.
+    ///
+    /// Return the command state unchanged if there is no appropriate substitute.
+    pub fn adapt_builtin_request(&self, command_state: GenericCommandState) -> GenericCommandState {
+        match self {
+            OperationAction::BuiltInAction(_, _) => {
+                command_state.update(GenericStateUpdate::scheduled())
+            }
+            OperationAction::AwaitBuiltInAction(_, _) => {
+                command_state.update(GenericStateUpdate::executing())
+            }
+            _ => command_state,
+        }
+    }
```

Comment:
thought: it's nonobvious how setting a scheduled state leads to an operation being spawned, execution state leads to it being awaited, and whether or not the state being set is for current command or a different one (sub-command)

At the beginning I was a bit confused why `<builtin-operation-name>` moves to scheduled and `await-<builtin-operation-name>` moves to scheduling, but, as the comment says, if idea is for `builtin-*` to trigger an operation and `await-builtin-*` to await the result, then it's not clear to me how this is eventually achieved.

And particularly, when we're doing a sub-workflow execution, e.g. an example from the documentation:

```toml
[trigger_config_update]
operation = "config_update"
input.tedgeUrl = "http://127.0.0.1:8000/tedge/file-transfer/example/config_update/mosquitto-1234"
input.type = "mosquitto"
on_exec = "waiting_for_config_update"

[waiting_for_config_update]
action = "await-operation-completion"
timeout_second = 600
on_timeout = "timeout_config_update"
on_success = "successful_config_update"
on_error = { status = "failed", reason = "fail to update the config"}
```

it's relatively clear that `action = "await-...` below is waiting for the entire config update to complete, but when applied to the lower level of states, I have trouble understanding why that's necessary for the executing state.

Reply:
> thought: it's nonobvious how setting a scheduled state leads to an operation being spawned, execution state leads to it being awaited, and whether or not the state being set is for current command or a different one (sub-command)

In both case, the workflow engine [delegates](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_agent/src/operation_workflows/message_box.rs#L19) the action to the [actor registered for that builtin operation](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_agent/src/operation_workflows/message_box.rs#L34).

The response is sent back by the operation specific actor to the workflow actor [for further processing](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_agent/src/operation_workflows/actor.rs#L79).
[This commit](https://github.com/thin-edge/thin-edge.io/pull/3070/commits/73dd8fd079ececa9edaa4f99feba9867d4e57c4c) improves method name as they were not really clear,

> It seems like there is an implicit assumption that `scheduling` state is handled by other agents trivially and immediately, so we model it as moving immediately to `executing` state, and we also assume that `executing` is a potentially long-running process that we need to await the completion of, but the question is, is this distinction necessary.

There is indeed an assumption that all the builtin operator actors follow the same workflow, transitioning from scheduled to executing and then either to successful or failed. However, there is no assumption on the timing. Technically, in both cases (scheduled and executing), the workflow engine awaits for the transition to occur.

Is this distinction necessary? From a workflow perspective, yes, as this makes explicit the transitions (todo - doing - done) and clarifies error management. This is notably a key point for the restart operation, where the device reboots between the two states. Internally, for the builtin actors, no. This is a legacy of when these actors where directly exposed to MQTT notifying themselves all the state transitions.

> In any case, I have now a bit better idea of the flow after looking at `adapt_builtin_request` and `adapt_builtin_response` methods:

> ```toml
> [scheduled]
>  action = "software_update"
>  on_exec = "executing"
>
>[executing]
>  action = "await_software_update"
>  on_success = "postprocess"
>  on_error = "rollback"
>```

> 1. workflow is in `scheduled` state, builtin operation actor is notified of that
> 2. builtin operation actor moves to `scheduled` state
> 3. builtin operation actor moves to `executing` state, tells workflow to run `on_exec` handler
> 4. workflow moves to `executing` state, builtin operation actor is notified of that
> 5. builtin operation actor moves to `successful` or `failed` state, tells workflow to run `on_success` or `on_error` handlers
> 6. workflow moes to `successful` or `failed` state
> 7. workflow is terminated

Your understanding is correct.